### PR TITLE
Allow custom metrics for DCGM Exporter

### LINKED ISSUE #934: [Feature] Support for new `customMetrics` value in DCGM Exporter
https://github.com/NVIDIA/dcgm-exporter/pull/351 adds support for user-definable custom metrics directly in the Helm values file which greatly simplifies the user experience needed when deploying DCGM Exporter with emission of custom metrics. Adding support for this field in the [operator's values file](https://github.com/NVIDIA/gpu-operator/blob/main/deployments/gpu-operator/values.yaml#L309) would be very useful. Without it, users would have to resort to the [(values-file undocumented) field in the ClusterPolicy](https://github.com/NVIDIA/gpu-operator/blob/main/deployments/gpu-operator/crds/nvidia.com_clusterpolicies_crd.yaml#L311-L319) at `config`, which accepts the name of a ConfigMap, meaning users have to pre-create this prior to deploying the operator (see also template ref [here](https://github.com/NVIDIA/gpu-operator/blob/main/deployments/gpu-operator/templates/clusterpolicy.yaml#L515)).

Please consider adding support for the new `customMetrics` field in the ClusterPolicy CRD and, by extension, in the Helm values of the operator to allow users to deploy the operator and DCGM Exporter with a list of custom metrics upfront.

### PR DESCRIPTION
**Pull request by chipzoller**
Closes #934 
Supersedes #935
Relates to NVIDIA/dcgm-exporter/pull/351

## Summary

This PR allows users to define a list of custom DCGM Exporter metrics during installation time. It secondarily documents in the values file the existence of the `config` element which, prior to this change, allowed users to point to an existing ConfigMap which contained such custom metrics. After this change users will be able to:

1. Use the values file as documentation to discover available fields for configuring DCGM Exporter via the Operator.
2. Define in-line to their values a list of their own custom metrics, obviating the need to pre-create a ConfigMap with the same contents.

This PR uses #803 by @frittentheke as a model to achieve as much consistency as possible for UX purposes.

## Demo

Templating results with defaults as shown in this PR.

<details><summary>Details</summary>
<p>

```sh
helm template gpu-operator . -s templates/clusterpolicy.yaml | yq .spec.dcgmExporter
```

```yaml
enabled: true
repository: nvcr.io/nvidia/k8s
image: dcgm-exporter
version: "3.3.7-3.5.0-ubuntu22.04"
imagePullPolicy: IfNotPresent
env:
  - name: DCGM_EXPORTER_LISTEN
    value: :9400
  - name: DCGM_EXPORTER_KUBERNETES
    value: "true"
  - name: DCGM_EXPORTER_COLLECTORS
    value: /etc/dcgm-exporter/dcp-metrics-included.csv
serviceMonitor:
  additionalLabels: {}
  enabled: false
  honorLabels: false
  interval: 15s
  relabelings: []
```

</p>
</details> 

Templating results with `dcgmExporter.config.name` defined. Same results with the addition of `create=false`.

<details><summary>Details</summary>
<p>

```sh
helm template gpu-operator . -s templates/clusterpolicy.yaml | yq .spec.dcgmExporter
```

```yaml
enabled: true
repository: nvcr.io/nvidia/k8s
image: dcgm-exporter
version: "3.3.7-3.5.0-ubuntu22.04"
imagePullPolicy: IfNotPresent
env:
  - name: DCGM_EXPORTER_LISTEN
    value: :9400
  - name: DCGM_EXPORTER_KUBERNETES
    value: "true"
  - name: DCGM_EXPORTER_COLLECTORS
    value: /etc/dcgm-exporter/dcp-metrics-included.csv
config:
  name: custom-dcgm-exporter-metrics
serviceMonitor:
  additionalLabels: {}
  enabled: false
  honorLabels: false
  interval: 15s
  relabelings: []
```

</p>
</details> 

Templating results with `name`, `create=true`, and `data` populated.

<details><summary>Details</summary>
<p>

```sh
helm template gpu-operator . -s templates/clusterpolicy.yaml | yq .spec.dcgmExporter
```

```yaml
enabled: true
repository: nvcr.io/nvidia/k8s
image: dcgm-exporter
version: "3.3.7-3.5.0-ubuntu22.04"
imagePullPolicy: IfNotPresent
env:
  - name: DCGM_EXPORTER_LISTEN
    value: :9400
  - name: DCGM_EXPORTER_KUBERNETES
    value: "true"
  - name: DCGM_EXPORTER_COLLECTORS
    value: /etc/dcgm-exporter/dcp-metrics-included.csv
config:
  name: custom-dcgm-exporter-metrics
serviceMonitor:
  additionalLabels: {}
  enabled: false
  honorLabels: false
  interval: 15s
  relabelings: []
```

```sh
helm template gpu-operator . -s templates/dcgm_exporter_config.yaml
```

```yaml
---
# Source: gpu-operator/templates/dcgm_exporter_config.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: custom-dcgm-exporter-metrics
  namespace: default
  labels:
    app.kubernetes.io/name: gpu-operator
    helm.sh/chart: gpu-operator-v24.6.1
    app.kubernetes.io/instance: gpu-operator
    app.kubernetes.io/version: "v24.6.1"
    app.kubernetes.io/managed-by: Helm
data:
  dcgm-metrics.csv: |
    # Format
    # If line starts with a '#' it is considered a comment
    # DCGM FIELD, Prometheus metric type, help message
    
    # Clocks
    DCGM_FI_DEV_SM_CLOCK,  gauge, SM clock frequency (in MHz).
    DCGM_FI_DEV_MEM_CLOCK, gauge, Memory clock frequency (in MHz).
```

Installing chart from this PR and verifying the DCGM Exporter contains the volume pointing to the new ConfigMap.

```sh
kubectl -n nvidia-gpu-operator get ds nvidia-dcgm-exporter -o yaml | yq .spec.template.spec.volumes
```

```yaml
- hostPath:
    path: /var/lib/kubelet/pod-resources
    type: ""
  name: pod-gpu-resources
- hostPath:
    path: /run/nvidia
    type: ""
  name: run-nvidia
- configMap:
    defaultMode: 420
    items:
      - key: dcgm-metrics.csv
        path: dcgm-metrics.csv
    name: custom-dcgm-exporter-metrics
  name: metrics-config
```

</p>
</details>

### CODE REVIEW

**Comment by chipzoller** [comment]
cc @mikemckiernan as this will probably require docs coverage.

**Review by cdesiniotis — APPROVED** [review]

**Comment by cdesiniotis** [comment]
@tariq1890 PTAL

File: deployments/gpu-operator/templates/clusterpolicy.yaml [hunk 1]
```diff
@@ -511,8 +511,9 @@ spec:
     {{- if .Values.dcgmExporter.args }}
     args: {{ toYaml .Values.dcgmExporter.args | nindent 6 }}
     {{- end }}
-    {{- if .Values.dcgmExporter.config }}
-    config: {{ toYaml .Values.dcgmExporter.config | nindent 6 }}
+    {{- if ((.Values.dcgmExporter.config).name) }}
```
Comment (tariq1890) [inline deployments/gpu-operator/templates/clusterpolicy.yaml]:
```suggestion
    {{- if  and (.Values.dcgmExporter.config) (.Values.dcgmExporter.config.name) }}
```

File: deployments/gpu-operator/templates/dcgm_exporter_config.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+metadata:
+  name: {{ .Values.dcgmExporter.config.name }}
+  namespace: {{ .Release.Namespace }}
+  labels:
+    {{- include "gpu-operator.labels" . | nindent 4 }}
+data:
+  dcgm-metrics.csv: |
+{{- .Values.dcgmExporter.config.data | nindent 4 }}
+{{- end }}
+{{- end }}
```
Comment (tariq1890) [inline deployments/gpu-operator/templates/dcgm_exporter_config.yaml]: add EOL

**Comment by tariq1890** [comment]
Hi @chipzoller , thank you for your contribution! I have left some comments. Can you also squash the commit history after addressing the comments?

File: deployments/gpu-operator/templates/clusterpolicy.yaml (hunk 1 above)
Comment (chipzoller) [reply deployments/gpu-operator/templates/clusterpolicy.yaml]: What I've written is basically the exact same effect but simpler. Do you still want this change?

File: deployments/gpu-operator/templates/dcgm_exporter_config.yaml (hunk 2 above)
Comment (chipzoller) [reply deployments/gpu-operator/templates/dcgm_exporter_config.yaml]: Not sure I follow you on this comment. Can you explain?

File: deployments/gpu-operator/templates/clusterpolicy.yaml (hunk 1 above)
Comment (tariq1890) [reply deployments/gpu-operator/templates/clusterpolicy.yaml]: Oh, is `if ((.Values.dcgmExporter.config).name)` the shorthand equivalent of the AND check I suggested above?

File: deployments/gpu-operator/templates/dcgm_exporter_config.yaml (hunk 2 above)
Comment (tariq1890) [reply deployments/gpu-operator/templates/dcgm_exporter_config.yaml]: If you look at the diff of this file, you'll see a red icon added by GitHub which denotes a missing EOL character. You can fix this by adding an empty newline which will act as the EOL marker of your file

See [here](https://stackoverflow.com/questions/2287967/why-is-it-recommended-to-have-empty-line-in-the-end-of-a-source-file) for context

File: deployments/gpu-operator/templates/clusterpolicy.yaml (hunk 1 above)
Comment (chipzoller) [reply deployments/gpu-operator/templates/clusterpolicy.yaml]: Yes

**Comment by chipzoller** [comment]
@tariq1890, can you just do a squash merge on this from your side?

**Comment by chipzoller** [comment]
Squash should be good now.

File: deployments/gpu-operator/templates/clusterpolicy.yaml (hunk 1 above)
Comment (cdesiniotis) [reply deployments/gpu-operator/templates/clusterpolicy.yaml]: @chipzoller can you point us to helm documentation that covers this shortened syntax? I would prefer Tariq's suggestion for better readability.

Comment (chipzoller) [reply deployments/gpu-operator/templates/clusterpolicy.yaml]: I don't know it's specific to Helm; it's just general Go templating and something I've learned over years of working with the tool. But I just changed it according to your suggestion.

Comment (cdesiniotis) [reply deployments/gpu-operator/templates/clusterpolicy.yaml]: Maybe I am having an issue viewing this in GitHub, but I do not see the change applied. I still see the shortened syntax:
```
{{- if ((.Values.dcgmExporter.config).name) }}
```

Comment (chipzoller) [reply deployments/gpu-operator/templates/clusterpolicy.yaml]: Do you not see it [here](https://github.com/NVIDIA/gpu-operator/pull/949/commits/ba4c9e1f1a6594d6598d797c7589f778410dbf99)?

Comment (cdesiniotis) [reply deployments/gpu-operator/templates/clusterpolicy.yaml]: The commit you linked is updating the wrong file.

Comment (chipzoller) [reply deployments/gpu-operator/templates/clusterpolicy.yaml]: 🤦 so sorry. Changed the wrong template. Should be updated.

File: deployments/gpu-operator/templates/clusterpolicy.yaml [hunk 3]
```diff
@@ -511,8 +511,9 @@ spec:
     {{- if .Values.dcgmExporter.args }}
     args: {{ toYaml .Values.dcgmExporter.args | nindent 6 }}
     {{- end }}
-    {{- if .Values.dcgmExporter.config }}
-    config: {{ toYaml .Values.dcgmExporter.config | nindent 6 }}
+    {{- if and (.Values.dcgmExporter.config) (.Values.dcgmExporter.config.name) }}
+    config:
+      name: {{ .Values.dcgmExporter.config.name }}
```
Comment (cdesiniotis) [inline deployments/gpu-operator/templates/clusterpolicy.yaml:516]: Aren't we forgetting to add `config.create` and `config.data`? Did you mean the below?
```suggestion
    config: {{ toYaml .Values.dcgmExporter.config | nindent 6 }}
```

**Review by cdesiniotis — COMMENTED** [review]
@chipzoller can you squash your commits again and update the commit message to describe the changes you are making? Thanks

File: deployments/gpu-operator/templates/clusterpolicy.yaml (hunk 3 above)
Comment (cdesiniotis) [reply deployments/gpu-operator/templates/clusterpolicy.yaml:516]: Oh disregard my comment, these are not relevant for clusterpolicy.

**Comment by chipzoller** [comment]
Done, @cdesiniotis.

**Review by tariq1890 — APPROVED** [review]

**Comment by chipzoller** [comment]
Looks like CI failing not related to the contents of this PR.

**Comment by tory34** [comment]
Why weren't these changes added to the latest release?

**Comment by chipzoller** [comment]
Probably because the latest release is a patch whereas this is a feature.

**Pull request merged**